File: iAsync.ObjC/lib/JFFStoreKit/JFFStoreKit/asyncAdapters/appStoreReceiptData.m
#import "appStoreReceiptData.h"

#import "NSError+CanceledPurchase.h"

static NSString *const mergeObject = @"2e94d46d-9f8f-4f6c-ac94-2ee1289b3c47";

@interface JFFAsyncAppStoreReceiptData : NSObject <
SKRequestDelegate,
JFFAsyncOperationInterface
>
@end

@implementation JFFAsyncAppStoreReceiptData
{
    SKReceiptRefreshRequest *_refreshReceiptRequest;
    JFFDidFinishAsyncOperationCallback _finishCallback;
}

- (void)unsubscribeFromObservervation
{
    _refreshReceiptRequest.delegate = nil;
    _refreshReceiptRequest = nil;
    _finishCallback        = nil;
}

- (void)asyncOperationWithResultCallback:(JFFDidFinishAsyncOperationCallback)finishCallback
                         handlerCallback:(JFFAsyncOperationChangeStateCallback)handlerCallback
                        progressCallback:(JFFAsyncOperationProgressCallback)progressCallback
{
    NSURL *receiptUrl = [[NSBundle mainBundle] appStoreReceiptURL];
    if ([[NSFileManager defaultManager] fileExistsAtPath:[receiptUrl path]]) {
        
        NSData *ios7ReceiptData = [[NSData alloc] initWithContentsOfURL:receiptUrl];
        if (finishCallback) {
            finishCallback(ios7ReceiptData, nil);
        }
    } else {
        
        _finishCallback = [finishCallback copy];
        
        _refreshReceiptRequest = [[SKReceiptRefreshRequest alloc] initWithReceiptProperties:@{}];
        _refreshReceiptRequest.delegate = self;
        [_refreshReceiptRequest start];
    }
}

- (void)doTask:(JFFAsyncOperationHandlerTask)task
{
    NSParameterAssert(task <= JFFAsyncOperationHandlerTaskCancel);
    
    if (task == JFFAsyncOperationHandlerTaskUnSubscribe) {
        [self unsubscribeFromObservervation];
    } else {
        [_refreshReceiptRequest cancel];
        _refreshReceiptRequest = nil;
    }
}

#pragma mark SKRequestDelegate

- (void)requestDidFinish:(SKRequest *)request
{
    if (_refreshReceiptRequest != request) {
        return;
    }
    
    JFFDidFinishAsyncOperationCallback finishCallback = _finishCallback;
    
    [self unsubscribeFromObservervation];
    
    NSURL *receiptUrl = [[NSBundle mainBundle] appStoreReceiptURL];
    if ([[NSFileManager defaultManager] fileExistsAtPath:[receiptUrl path]]) {
        
        if (finishCallback) {
            finishCallback([[NSData alloc] initWithContentsOfURL:receiptUrl], nil);
        }
    } else {
        
        if (finishCallback) {
            finishCallback(nil, [JFFSilentError newErrorWithDescription:@"no receipt was recieved"]);
        }
    }
}

- (void)request:(SKRequest *)request didFailWithError:(NSError *)error
{
    if (_refreshReceiptRequest != request) {
        return;
    }
    
    JFFDidFinishAsyncOperationCallback finishCallback = _finishCallback;
    
    [self unsubscribeFromObservervation];
    
    if (finishCallback) {
        if ([error isCanceledPurchaseAuthorization]) {
            error = [JFFAsyncOpFinishedByCancellationError new];
        }
        finishCallback(nil, error);
    }
}

@end

JFFAsyncOperation appStoreReceiptDataLoader(void)
{
    JFFAsyncOperationInstanceBuilder factory = ^id<JFFAsyncOperationInterface>(void) {
        return [JFFAsyncAppStoreReceiptData new];
    };
    
    JFFAsyncOperation loader = buildAsyncOperationWithAdapterFactory(factory);
    
    return [mergeObject asyncOperationMergeLoaders:loader withArgument:@(__FUNCTION__)];
}

File: iAsync.Swift/lib/JStoreKit/JStoreKit/asyncAdapters/appStoreReceiptData.swift
//
//  appStoreReceiptData.swift
//  JStoreKit
//
//  Created by Vladimir Gorbenko on 03.10.14.
//  Copyright (c) 2014 EmbeddedSources. All rights reserved.
//

import Foundation

import JAsync
import JUtils
import StoreKit

private extension NSError {

    var isCanceledPurchaseAuthorization: Bool {
        let result = "SSErrorDomain" == domain && code == 16
        return result
    }
}

private let cachedAsyncOp = JCachedAsync<String, NSData>()

private class JAsyncAppStoreReceiptData : NSObject, SKRequestDelegate, JAsyncInterface {
    
    typealias ResultType = NSData
    
    private var refreshReceiptRequest: SKReceiptRefreshRequest?
    private var finishCallback: JAsyncTypes<ResultType>.JDidFinishAsyncCallback?
    
    func unsubscribeFromObservervation() {
        
        if let refreshReceiptRequest = refreshReceiptRequest {
            
            refreshReceiptRequest.delegate = nil
            self.refreshReceiptRequest = nil
        }
        finishCallback = nil
    }
    
    private var receiptData: NSData? {
        if let receiptUrl = NSBundle.mainBundle().appStoreReceiptURL {
            return NSData(contentsOfURL:receiptUrl)
        }
        return nil
    }
    
    func asyncWithResultCallback(
        finishCallback: JAsyncTypes<ResultType>.JDidFinishAsyncCallback,
        stateCallback: JAsyncChangeStateCallback,
        progressCallback: JAsyncProgressCallback)
    {
        if let receiptData = receiptData {
        
            finishCallback(result: JResult.value(receiptData))
        } else {
        
            self.finishCallback = finishCallback
        
            let refreshReceiptRequest  = SKReceiptRefreshRequest(receiptProperties:[:])
            self.refreshReceiptRequest = refreshReceiptRequest
            refreshReceiptRequest.delegate = self
            refreshReceiptRequest.start()
        }
    }
    
    func doTask(task: JAsyncHandlerTask) {
        
        assert(task.rawValue <= JAsyncHandlerTask.Cancel.rawValue)
        
        if task == .UnSubscribe {
            unsubscribeFromObservervation()
        } else {
            refreshReceiptRequest?.cancel()
            refreshReceiptRequest = nil
        }
    }

    var isForeignThreadResultCallback: Bool {
        return false
    }
    
    func requestDidFinish(request: SKRequest!) {
        
        if refreshReceiptRequest != request {
            return
        }
    
        let finishCallback = self.finishCallback
    
        unsubscribeFromObservervation()
    
        if let receiptData = receiptData {
        
            finishCallback?(result: JResult.value(receiptData))
        } else {
        
            finishCallback?(result: JResult.error(JSilentError(description:"no receipt was recieved")))
        }
    }
    
    func request(request: SKRequest!, didFailWithError error: NSError!) {
        
        if refreshReceiptRequest != request {
            return
        }
        
        let finishCallback = self.finishCallback
        
        unsubscribeFromObservervation()
        
        if let finishCallback = finishCallback {
            if error.isCanceledPurchaseAuthorization {
                
                finishCallback(result: JResult.error(JAsyncFinishedByCancellationError()))
                return
            }
            finishCallback(result: JResult.error(error))
        }
    }
}

func appStoreReceiptDataLoader() -> JAsyncTypes<NSData>.JAsync {
    
    let factory = { () -> JAsyncAppStoreReceiptData in
        return JAsyncAppStoreReceiptData()
    }
    
    let loader = JAsyncBuilder.buildWithAdapterFactory(factory)
    
    return cachedAsyncOp.asyncOpMerger(loader, uniqueKey:__FUNCTION__)
}
